accounts/models.py

```diff
@@ -350,55 +381,123 @@ def num_sounds_pending_moderation(self):
+        Even in this case a DeletedUser object will be created to keep a record.
+
+        Args:
+            remove_sounds (bool): if True the sounds created by the user will be deleted as well. Otherwise the sounds
+              will still be available to other users but appear under a "deleted user" account. Defaults to False.
+            delete_user_object_from_db (bool): if True the user object will be completely removed from the DB together
+              with all related content. Defaults to False.
+            deletion_reason (str): reason for the user being deleted. Should be one of the choices defined in
+              DeletedUser.DELETION_REASON_CHOICES. Defaults to DeletedUser.DELETION_REASON_DELETED_BY_ADMIN.
         """
 
-        self.user.username = 'deleted_user_%s' % self.user.id
-        self.user.email = 'deleted_user_%s@freesound.org' % self.user.id
-        self.has_avatar = False
-        self.is_deleted_user = True
-        self.user.set_unusable_password()
-
-        self.about = ''
-        self.home_page = ''
-        self.signature = ''
-        self.geotag = None
-
-        self.save()
-        self.user.save()
-        if remove_sounds:
-            Sound.objects.filter(user=self.user).delete()
-            Pack.objects.filter(user=self.user).update(is_deleted=True)
-        else:
-            Sound.objects.filter(user=self.user).update(is_index_dirty=True)
+        # Run all deletion operations in a single transaction so if there is an error we don't create duplicate
+        # DeletedUser objects
+        with transaction.atomic():
+
+            # Create a DeletedUser object to store basic information for the record (first check if it
+            # already exists because user was deleted previously but db object preserved
+            try:
+                deleted_user_object = DeletedUser.objects.get(user=self.user)
+            except DeletedUser.DoesNotExist:
+                deleted_user_object = DeletedUser.objects.create(
+                    user=self.user,
+                    username=self.user.username,
+                    email=self.user.email,
+                    date_joined=self.user.date_joined,
+                    last_login=self.user.last_login,
+                    reason=deletion_reason)
+
+            # If UserGDPRDeletionRequest object(s) exist for that user, update their status and deleted_user property
+            UserGDPRDeletionRequest.objects.filter(user_id=self.user.id)\
+                .update(status="de", deleted_user=deleted_user_object)
+
+            # If Sound/Pack objects will be deleted, we need to make sure that num_sound_downloads and
+            # num_pack_downloads fileds from users that downloaded packs or sounds from the user being delted are
+            # correct. Otherwise this will make the deletion process to fail (violating DB constraints)
+            if delete_user_object_from_db or remove_sounds:
```

Comment:
If this is a popular sound, then we'll perform a lot of queries here, 2 per downloader - one to get the count and one to save it. If a user has download a sound twice, we'll process it twice.
We should
 * do a unique select when getting the downloaders
 * Do only 1 query to get all the download counts (i.e. `user_id__in=[sound_downloaders]`), to minimise the number of queries that we make here

Reply:
didn't see a reply from you about this

Reply:
Addressed in c654917. Now we use PG's Greatest to make sure the count fields are not set to < 0 and therefore we avoid the need of updating the fields before deletion.